### PR TITLE
Checkout service

### PR DESCRIPTION
**Pull request by melnikovi**

### CODE REVIEW

**Comment by buskamuza** [comment]
Please update the main document with a reference to this one, so it's possible to find it.

File: design-documents/checkout-service.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+### Place Order Operation
+
+Communication for place order operation looks similar. The only major differences are that we need to save payment, load quote multiple times and save quote. This flow potentially can be optimized.
+
+
+![Apply more decomposition](checkout-service/place-order.png)
+
+
+## Transactions
+Currently Quote module is responsible for creating an order. Need to move this responsibility to order service (BFF after first iteration).
```
Comment (buskamuza) [inline design-documents/checkout-service.md]: "BFF" - do you mean "monolith"? Thought it's the same thing after first iteration, the meaning is different. It's strange responsibility for BFF to place an order. Monolith represents services, just deployed on a single instance (with in-memory dependencies), and in this case placing order is an OK responsibility for the monolith.

File: design-documents/checkout-service.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+BFF will be responsible for loading quote and managing the transaction. If order placing failed, no additional transactions need to be rolled back after first iteration as all transactions being performed on BFF. If order placing successful, but  removing quote failed, we would need to retry removing quote instead of rolling back. Mechanism of retrying of failed operations is a general concern and should be discussed separately.
+
+
+## Reporting
+For reporting purposes, service should data to reporting service (BFF after first iteration).
```
Comment (buskamuza) [inline design-documents/checkout-service.md]: Also, a separate generic question.

"should data" - "should send data"?

File: design-documents/checkout-service.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+
+## Open Questions
+* Format of the data API receives is different from the format service contracts use in some cases
+* Hydration/extraction for certain entities triggers initialization of resource models/queries to database, ProductInterface is one of the examples
+* Code reuse (reusing the same modules). It's anti pattern, but maybe for us it's ok?
```
Comment (buskamuza) [inline design-documents/checkout-service.md]: Modules or framework/libraries? What are the examples of modules?

File: design-documents/checkout-service.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
+# Checkout Service
+
+## Overview
+
+This document describes design and communication of monolith application and checkout service.
+
+## Design
+
+Service should have it's own core_config_data table that would contain settings specific to the service. Service should expose API to modify settings in core_config_data. Admin panel should support editing settings in the local database (settings for BFF) and for individual services via remote calls.
```
[the remark this answers is not in the document]
Comment (melnikovi) [reply design-documents/checkout-service.md]: Would you suggest add this to service isolation document or to the new one?

File: design-documents/checkout-service.md (hunk 1 above)
Comment (melnikovi) [reply design-documents/checkout-service.md]: Yes, by BFF I mean monolith. I specified that placing of an order is responsibility of order service. Are you suggesting to reword it?

File: design-documents/checkout-service.md (hunk 2 above)
Comment (melnikovi) [reply design-documents/checkout-service.md]: Fixed the sentence and moved to [separate document](https://github.com/magento/architecture/pull/60).

File: design-documents/checkout-service.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+Service should not have knowledge about stores/websites. However as service stores information about stores/websites within the quote stores/websites need to be validated. Service can request information about stores/website to use use for validation. Or validation can be omitted (recommended).
+
+Directory data (countries, cities, states) should be moved to configuration (preferable) or moved to a separate service/exposed on BFF.
+
+Operations that are currently per item (inventory requests) need to be done for multiple items to reduce number of calls.
+
+Dependency on session should be removed. All methods should receive customer with customer groups as an argument.
```
[the remark this answers is not in the document]
Comment (melnikovi) [reply design-documents/checkout-service.md]: Moved to a [separate document](https://github.com/magento/architecture/pull/60).

File: design-documents/checkout-service.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+This document describes design and communication of monolith application and checkout service.
+
+## Design
+
+Service should have it's own core_config_data table that would contain settings specific to the service. Service should expose API to modify settings in core_config_data. Admin panel should support editing settings in the local database (settings for BFF) and for individual services via remote calls.
+
+Service should not have knowledge about stores/websites. However as service stores information about stores/websites within the quote stores/websites need to be validated. Service can request information about stores/website to use use for validation. Or validation can be omitted (recommended).
+
+Directory data (countries, cities, states) should be moved to configuration (preferable) or moved to a separate service/exposed on BFF.
```
[the remark this answers is not in the document]
Comment (melnikovi) [reply design-documents/checkout-service.md]: Added explanation and moved to a [separate document](https://github.com/magento/architecture/pull/60).

File: design-documents/checkout-service.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+Operations that are currently per item (inventory requests) need to be done for multiple items to reduce number of calls.
+
+Dependency on session should be removed. All methods should receive customer with customer groups as an argument.
+
+Shared interfaces (product interfaces) need to be moved into API modules. For backwards compatibility we can export interfaces in the new API modules under old name. Shared configuration (configuration for extension attributes) need to be moved to a separate modules as well.
+
+BFF should expose all quote (and other services we separate from monolith) API and work like proxy for all calls to maintain backwards compatibility for web API.
```
[the remark this answers is not in the document]
Comment (melnikovi) [reply design-documents/checkout-service.md]: Reworded and moved to a [separate document](https://github.com/magento/architecture/pull/60).

File: design-documents/checkout-service.md (hunk 3 above)
Comment (melnikovi) [reply design-documents/checkout-service.md]: Modules that contain definitions of extension attributes, interfaces or implementations of the interfaces that enable over network communication.

File: design-documents/checkout-service.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+## Overview
+
+This document describes design and communication of monolith application and checkout service.
+
+## Design
+
+Service should have it's own core_config_data table that would contain settings specific to the service. Service should expose API to modify settings in core_config_data. Admin panel should support editing settings in the local database (settings for BFF) and for individual services via remote calls.
+
+Service should not have knowledge about stores/websites. However as service stores information about stores/websites within the quote stores/websites need to be validated. Service can request information about stores/website to use use for validation. Or validation can be omitted (recommended).
```
[the remark this answers is not in the document]
Comment (melnikovi) [reply design-documents/checkout-service.md]: Added explanation and moved to a [separate document](https://github.com/magento/architecture/pull/60).

**Pull request merged**